Add RESP2 to RESP3 replies migration guide (#2513)
redis/redis-doc: docs/reference/resp2-to-resp3-replies.md

---
title: "RESP2 to RESP3 replies migration guide"
linkTitle: "RESP2 to RESP3 migration"
weight: 4
description: RESP2 to RESP3 replies reference for clients developers
aliases:
    - /topics/resp2-to-resp3-replies/
---

> In the journey from RESP2 to [RESP3](https://redis.io/docs/reference/protocol-spec/), one of the primary motivations was 

Reviewer: IMHO we shouldn't do any manual line breaking, but rather break at periods, see the README in the repo

Author: Fixed

to streamline the developer experience by simplifying response parsing. Over time, it became evident that developers 
frequently performed specific data transformations on RESP2 replies. Addressing these transformations directly at the 

> protocol level not only reduces the overhead for developers but also ensures a more consistent and efficient data 

Reviewer: does this means to tell the story of why RESP3 was created?
i think it should mention something about replies standing for themselves, without the need for the client library to know which command generated them.

Author: Updated. Thanks!

handling process. This documentation provides a reference for developers to migrate their clients from RESP2 to RESP3.

### Command Replies Comparison
- If you see any missing commands, please [open an issue](https://github.com/redis/redis-doc/issues/new?title=RESP2-to-RESP3%20command%20is%20missing)

> - `null` was "promoted" from a subtype of `BlobString` and `Array` to its own type, these changes are not included in this list.

Reviewer: is this suppose to represent a general list of changes between RESP2 and RESP2?
in which case many lines are missing (double, boolean, map, etc).
i don't think we should repeat it, but rather add a prompt line that redirects to the spec.

instead, since the tables below only describe commands with non-trivial changes, this block could be a prompt statement about generic transformation, like an interlaced array becoming a map, and clearly stating that the list below covers only the non-trivial changes.

alternatively, if we want, we can put here a full list of the common transformation (like interlaced array to map, string to double, int to boolean), i.e. anything that can generically be applied, so that we don't need to list all the affected commands.

Author: Good point. Updated

| ZUNION WITHSCORES        | Array\<BlobString>                                          | Array<\[BlobString, Double]>                            |
| ZRANK WITHSCORE          | \[number, BlobString]                                       | \[number, Double]                                       |
| ZREVRANK WITHSCORE       | \[number, BlobString]                                       | \[number, Double]                                       |
| ZRANGE WITHSCORES        | Array\<BlobString>                                          | Array<\[BlobString, Double]                             |
| ZRANGEBYSCORE WITHSCORES | Array\<BlobString>                                          | Array<\[BlobString, Double]                             |
| ZMPOP                    | \[BlobString, Array<\[BlobString, BlobString]>]             | \[BlobString, Array<\[BlobString, Double]>]             |
| BZPOPMAX                 | null &#124; \[BlobString, BlobString, BlobString]           | null &#124; \[BlobString, BlobString, Double]           |
| BZPOPMIN                 | null &#124; \[BlobString, BlobString, BlobString]           | null &#124; \[BlobString, BlobString, Double]           |

> | ZPOPMIN                  | null &#124; \[BlobString, BlobString]                       | null &#124; \[BlobString, Double]                       |

Reviewer 2: sorry to be pedantic, but can we be super clear about what this syntax means? i.e. `Array<...>` vs `[...]` ? it *looks* like you're using `[x,y,z]` to mean an array of exactly these elements, and `Array<...>` to mean an array of some unknown quantity of elements; is that right? so `Array<\[x,y]>` is an array (unknown length) of jagged pairs?

Author: Added to the doc.

---

> title: "RESP2 to RESP3 replies migration guide"

Reviewer 3: ```suggestion
title: "RESP2 to RESP3 reply migration guide"
```

linkTitle: "RESP2 to RESP3 migration"
weight: 4

> description: RESP2 to RESP3 replies reference for clients developers

Reviewer 3: ```suggestion
description: RESP2 to RESP3 reply reference for client library developers
```

aliases:
    - /topics/resp2-to-resp3-replies/
---

> In the journey from RESP2 to [RESP3](https://redis.io/docs/reference/protocol-spec/), one of the primary motivations was to streamline the developer experience by simplifying response parsing. 
> Over time, it became evident that developers frequently performed specific data transformations on RESP2 replies, which required hardcoded transformations per command. 
> Addressing these transformations directly at the protocol level reduces clients' overhead and ensures a more consistent and efficient replies handling process.  
> This documentation provides a reference for developers to migrate their clients from RESP2 to RESP3.

Reviewer 3: ```suggestion
The primary motivation for creating the [RESP3](https://redis.io/docs/reference/protocol-spec/) protocol, the successor to RESP2, was to streamline the developer experience by simplifying response parsing. 
It became evident that developers frequently performed specific data transformations on RESP2 replies, which required hardcoded transformations for each command. 
Addressing these transformations directly at the protocol level reduces client overhead and ensures a more consistent and efficient reply-handling process.  
This documentation provides a reference guide to help developers migrate their clients from RESP2 to RESP3.

**Note**:
> Each of the Redis base command manual pages now includes both RESP2 and RESP3 responses.
```

> ### Command Replies Comparison
> - RESP3 introduces many new [simple types and aggregates]((https://redis.io/docs/reference/protocol-spec/#resp-protocol-description)).
> The following tables compare only commands with non-trivial changes in replies, primarily for new aggregates introduced in RESP3 that require major changes in the client implementation.

Reviewer 3: ```suggestion
### Command replies comparison
- RESP3 introduces many new [simple and aggregate reply types]((https://redis.io/docs/reference/protocol-spec/#resp-protocol-description)).
The following tables compare only commands with non-trivial changes to their replies, primarily for new aggregates introduced in RESP3 that require major changes in the client implementation.
```

- The types are described using [“TypeScript like” syntax](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html): 

>   - `[a, b]` stands for [a tuple](https://www.typescriptlang.org/docs/handbook/2/objects.html#tuple-types) where we know the exact number of elements and types at specific positions.
>   - `Array<a>` stands for [an array](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#arrays) where we know the type of elements but not the number of elements.
> 
> ### Cluster Management

Reviewer 3: ```suggestion
  - `[a, b]` stands for [a tuple](https://www.typescriptlang.org/docs/handbook/2/objects.html#tuple-types) where the exact number of elements and types at specific positions are known.
  - `Array<a>` stands for [an array](https://www.typescriptlang.org/docs/handbook/2/everyday-types.html#arrays) where the type of elements is known but not the number of elements.

### Cluster management
```

### Connection management
| Command     | RESP2 Reply | RESP3 Reply    |
|-------------|-------------|----------------|
| CLIENT INFO | BlobString  | VerbatimString |
| CLIENT LIST | BlobString  | VerbatimString |

### Generic
| Command   | RESP2 Reply | RESP3 Reply            |
|-----------|-------------|------------------------|
| RANDOMKEY | BlobString  | null &#124; BlobString |

### Geo
| Command                     | RESP2 Reply        | RESP3 Reply    |
|-----------------------------|--------------------|----------------|
| GEORADIUS WITHCOORD         | \[..., BlobString] | \[..., Double] |
| GEORADIUSBYMEMBER WITHCOORD | \[..., BlobString] | \[..., Double] |
| GEOSEARCH WITHCOORD         | \[..., BlobString] | \[..., Double] |

### Hash
| Command              | RESP2 Reply        | RESP3 Reply                      |
|----------------------|--------------------|----------------------------------|
| HGETALL              | Array              | Map                              |
| HRANDFIELD WITHVALUE | Array\<BlobString> | Array<\[BlobString, BlobString]> |

> ### Scripting and Functions

Reviewer 3: ```suggestion
### Scripting and functions
```

| FUNCTION LIST          | Array<\[‘library\_name’, BlobString, ‘engine’, BlobString, ‘functions’, Array<\[‘name’, BlobString, ‘description’, BlobString &#124; null, ‘flags’, Array\<BlobString>]>]>                              | Array<\[{ library\_name: BlobString, engine: BlobString, functions: Array<{ name: BlobString, description: BlobString &#124; null, flags: Set\<BlobString> }> }]>                            |
| FUNCTION LIST WITHCODE | Array<\[‘library\_name’, BlobString, ‘engine’, BlobString, ‘functions’, Array<\[‘name’, BlobString, ‘description’, BlobString &#124; null, ‘flags’, Array\<BlobString>]>, ‘library\_code’, BlobString]> | Array<\[{ library\_name: BlobString, engine: BlobString, functions: Array<{ name: BlobString, description: BlobString &#124; null, flags: Set\<BlobString> }>, library\_code: BlobString }]> |

> ### Server Management

Reviewer 3: ```suggestion
### Server management
```

| COMMAND       | Array<\[BlobString, number, Set\<BlobString>, number, number, number, Set\<BlobString>, Set\<BlobString>,]> |                                              |
| INFO          | BlobString                                                                                                  | VerbatimString                               |
| MEMORY DOCTOR | BlobString                                                                                                  | VerbatimString                               |

### Sentinel
| Command             | RESP2 Reply | RESP3 Reply |
|---------------------|-------------|-------------|
| SENTINEL CONFIG GET | Array       | Map         |

### String
| Command | RESP2 Reply | RESP3 Reply |
|---------|-------------|-------------|
| LCS IDX | Array       | Map         |

### Set
| Command                                                              | RESP2 Reply | RESP3 Reply |
|----------------------------------------------------------------------|-------------|-------------|
| SINTER <br/>SPOP COUNT<br/>SMEMBERS<br/>SRANDMEMBER COUNT<br/>SUNION | Array       | Set         |

> ### Sorted Set

Reviewer 3: ```suggestion
### Sorted set
```

---
title: "RESP2 to RESP3 replies migration guide"

> linkTitle: "RESP2 to RESP3 replies migration guide"

Reviewer 4: This needs to be shorted as it appears as the title in the left-hand navigation tree.
